modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -1485,69 +1485,89 @@ private void prepareIndexRebuildFuture(int cacheId) {
     }
 
     /** {@inheritDoc} */
-    @Override public void rebuildIndexesIfNeeded(GridDhtPartitionsExchangeFuture fut) {
+    @Override public void rebuildIndexesIfNeeded(GridDhtPartitionsExchangeFuture exchangeFut) {
+        assert nonNull(exchangeFut);
+
         GridQueryProcessor qryProc = cctx.kernalContext().query();
 
-        GridCompoundFuture compoundAllIdxsRebuilt = null;
+        if (!qryProc.moduleEnabled())
+            return;
 
-        if (qryProc.moduleEnabled()) {
-            for (final GridCacheContext cacheCtx : (Collection<GridCacheContext>)cctx.cacheContexts()) {
-                if (cacheCtx.startTopologyVersion().equals(fut.initialVersion())) {
-                    final int cacheId = cacheCtx.cacheId();
-                    final GridFutureAdapter<Void> usrFut = idxRebuildFuts.get(cacheId);
+        GridCompoundFuture allCacheIdxsCompoundFut = null;
 
-                    IgniteInternalFuture<?> rebuildFut = qryProc.rebuildIndexesFromHash(cacheCtx);
+        ExecutorService rebuildIdxExecSvc = null;
 
-                    if (rebuildFut != null) {
-                        log().info("Started indexes rebuilding for cache [name=" + cacheCtx.name()
-                            + ", grpName=" + cacheCtx.group().name() + ']');
+        for (GridCacheContext cacheCtx : (Collection<GridCacheContext>)cctx.cacheContexts()) {
+            if (!cacheCtx.startTopologyVersion().equals(exchangeFut.initialVersion()))
+                continue;
 
-                        if (compoundAllIdxsRebuilt == null)
-                            compoundAllIdxsRebuilt = new GridCompoundFuture();
+            int cacheId = cacheCtx.cacheId();
+            GridFutureAdapter<Void> usrFut = idxRebuildFuts.get(cacheId);
 
-                        compoundAllIdxsRebuilt.add(rebuildFut);
+            if (isNull(rebuildIdxExecSvc))
+                rebuildIdxExecSvc = qryProc.getIndexing().rebuildIndexExecutorService(0);
 
-                        rebuildFut.listen(new CI1<IgniteInternalFuture>() {
-                            @Override public void apply(IgniteInternalFuture fut) {
-                                idxRebuildFuts.remove(cacheId, usrFut);
+            IgniteInternalFuture<?> rebuildFut = qryProc.rebuildIndexesFromHash(cacheCtx, rebuildIdxExecSvc);
```

Comment:
Why rebuildFut is necessary ? Looks like usrFut can be used instead.

Reply:
Method [1] and [2] stores GridFutureAdapter. do you think it's better to store GridCompoundFuture in the [2]?

[1] - org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing#rebuildIndexesFromHash returns GridCompoundFuture
[2] - org.apache.ignite.internal.processors.cache.persistence.GridCacheDatabaseSharedManager#idxRebuildFuts